### PR TITLE
Make Periodic Table widget toggleable

### LINKED ISSUE #201: Do not show the periodic table by default
The latest version of the client shows the periodic table by default which I find really overwhelming and reduces the UI quality significantly. Embedded into the QE app I can't even see the whether a given query yielded any results because the screen is completely covered by the table (see screenshot).
Screen Shot 2020-11-03 at 3 27 07 PM
**Proposal**: Make the table an accordion widget that can be opened if desired.

### PR DESCRIPTION
**Pull request by CasperWA**
Closes #201.

See on [binder](https://mybinder.org/v2/gh/CasperWA/voila-optimade-client/close_201_toggle-periodic-table?urlpath=%2Fvoila%2Frender%2FOPTIMADE-Client.ipynb).

Wrap the checkbox and periodic table widget in a `VBox` container and add a button to change toggle the `visibility` layout parameter between `visible` and `hidden`.

A suggestion from my part would be to minimize the button and/or move it to the left side, but it might also be all right as is?

### CODE REVIEW

**Comment by csadorf** [comment]
Before I review this, the binder appears to be broken:
```
ModuleNotFoundError: No module named 'appmode'
Removing intermediate container 24295b9f96f2
The command '/bin/sh -c ./binder/postBuild' returned a non-zero code: 1
```

**Comment by CasperWA** [comment]
> Before I review this, the binder appears to be broken:
> 
> ```
> ModuleNotFoundError: No module named 'appmode'
> Removing intermediate container 24295b9f96f2
> The command '/bin/sh -c ./binder/postBuild' returned a non-zero code: 1
> ```

Fixed in #205. The link seems to work now.
To test the toggle button, simply choose a database to enable the button.

File: optimade_client/subwidgets/periodic_table.py [hunk 1]
```diff
@@ -11,20 +12,39 @@
 class PeriodicTable(ipw.VBox):
     """Wrapper-widget for PTableWidget"""
 
-    def __init__(self, **kwargs):
+    def __init__(self, show_widget: bool = True, **kwargs):
         self._disabled = kwargs.get("disabled", False)
 
+        self.toggle_button = ipw.ToggleButton(
+            value=show_widget,
+            description="Toggle Periodic Table",
```
Comment (csadorf) [inline optimade_client/subwidgets/periodic_table.py]:
```suggestion
            description="Show Periodic Table",
```

File: optimade_client/subwidgets/periodic_table.py [hunk 2]
```diff
@@ -67,3 +89,18 @@ def freeze(self):
     def unfreeze(self):
         """Activate widget (in its current state)"""
         self.disabled = False
+
+    def _toggle_widget(self, change: dict):
```
Comment (csadorf) [inline optimade_client/subwidgets/periodic_table.py:90]: Whether the table is shown or not should be determined by a widget trait "show_periodic_table_of_elements" and that trait is linked with the button.

File: optimade_client/query_filter.py [hunk 3]
```diff
@@ -81,7 +76,7 @@ def __init__(
         self.filter_header = ipw.HTML(
             '<h4 style="margin:0px;padding:0px;">Apply filters</h4>'
         )
-        self.filters = FilterTabs()
+        self.filters = FilterTabs(show_large_filters=not embedded)
```
Comment (csadorf) [inline optimade_client/query_filter.py:79]: How is this related to this PR?

File: optimade_client/subwidgets/periodic_table.py [hunk 4]
```diff
@@ -11,20 +12,39 @@
 class PeriodicTable(ipw.VBox):
     """Wrapper-widget for PTableWidget"""
 
-    def __init__(self, **kwargs):
+    def __init__(self, show_widget: bool = True, **kwargs):
```
Comment (csadorf) [inline optimade_client/subwidgets/periodic_table.py]: I think it would be better to make the `show_widget` constructor argument a a trait called `expanded` or so.

**Review by csadorf — DISMISSED** [review]
If the periodic table can only be toggled if a data base is selected that would defeat the purpose. So at least programmatically it should be possible to toggle the visibility independently. For this I suggest to add a trait to the widget and connect the toggle button internally with that trait. See also in-code comments for details.

Two additional points which are out of scope for this PR:
1. Whether the table is available/visible or not, I would expect to be able to always also enter elements "manually" into a text field which would be synchronized with the table.
2. It seems to me that there is no clear explanation as to what the various "selection states" represent as part of the query. Clicking elements multiple times makes them change color, and I assume based on the color that it means "included/excluded", but it is not immediately obvious that this is the meaning in the context of the query and it would be completely impossible to discern for a person with red/green color-blindness.

**Comment by CasperWA** [comment]
> If the periodic table can only be toggled if a data base is selected that would defeat the purpose. So at least programmatically it should be possible to toggle the visibility independently. For this I suggest to add a trait to the widget and connect the toggle button internally with that trait. See also in-code comments for details.

If `embedded=True` is passed to the `OptimadeQueryFilterWidget` widget the Periodic Table widget will be toggled "off"/hidden by default.
The current UX is to keep all filter inputs disabled until a database has been chosen. This is to guide the user to this point first.
Now to the issue here: Whether one can show or hide the disabled Periodic Table widget I thought to fall under the same umbrella, and the initial configuration can be set up using the `embedded` parameter. But I have no issue making the toggle button special in this regard. So if you think it seems more natural to be able to toggle the input, no matter if the shown/hidden widget is disabled or not, then I'll implement that here.

> 1. Whether the table is available/visible or not, I would expect to be able to always also enter elements "manually" into a text field which would be synchronized with the table.

This is a conscious UX choice at this point (to only use the Periodic Table widget).
However, I will raise this point as an issue, but I don't expect this alternative to be implemented in the near future.

> 2. It seems to me that there is no clear explanation as to what the various "selection states" represent as part of the query. Clicking elements multiple times makes them change color, and I assume based on the color that it means "included/excluded", but it is not immediately obvious that this is the meaning in the context of the query and it would be completely impossible to discern for a person with red/green color-blindness.

This is already the subject of issue #160.
The color-blindness is indeed a valid concern I hadn't previously considered.

File: optimade_client/query_filter.py (hunk 3 above)
Comment (CasperWA) [reply optimade_client/query_filter.py:79]: This is the inner implementation of the `embedded` parameter for `OptimadeQueryFilterWidget`.

File: optimade_client/subwidgets/periodic_table.py (hunk 1 above)
Comment (CasperWA) [reply optimade_client/subwidgets/periodic_table.py]: And then you want it to change description depending on the widget being shown/hidden? I put this in the tool tip instead of the description, but if you think it's better for the user to have not only the visual aid of the "pressed in" and "pressed out" button, but also the written change, then that could easily be implemented.

**Comment by csadorf** [comment]
> > If the periodic table can only be toggled if a data base is selected that would defeat the purpose. So at least programmatically it should be possible to toggle the visibility independently. For this I suggest to add a trait to the widget and connect the toggle button internally with that trait. See also in-code comments for details.
> 
> If `embedded=True` is passed to the `OptimadeQueryFilterWidget` widget the Periodic Table widget will be toggled "off"/hidden by default.
> The current UX is to keep all filter inputs disabled until a database has been chosen. This is to guide the user to this point first.
> Now to the issue here: Whether one can show or hide the disabled Periodic Table widget I thought to fall under the same umbrella, and the initial configuration can be set up using the `embedded` parameter. But I have no issue making the toggle button special in this regard. So if you think it seems more natural to be able to toggle the input, no matter if the shown/hidden widget is disabled or not, then I'll implement that here.
> 

I think that the API I'm suggesting provides more control and flexibility, but as long as the widget is hidden by default, it's fine with me. 

> > 1. Whether the table is available/visible or not, I would expect to be able to always also enter elements "manually" into a text field which would be synchronized with the table.
> 
> This is a conscious UX choice at this point (to only use the Periodic Table widget).
> However, I will raise this point as an issue, but I don't expect this alternative to be implemented in the near future.
> 

It feels non-intuitive and unnecessarily restrictive to me, but again, out of scope for this PR anyways.

**Comment by csadorf** [comment]
@CasperWA Is this ready for another review?

File: optimade_client/subwidgets/periodic_table.py (hunk 2 above)
Comment (CasperWA) [reply optimade_client/subwidgets/periodic_table.py:90]: Opened issue to track this suggestion #209

File: optimade_client/subwidgets/periodic_table.py [hunk 5]
```diff
@@ -11,20 +12,38 @@
...
-            disabled=kwargs.get("disabled", False),
+            disabled=self.disabled,
         )
         self.ptable = PTableWidget(**kwargs)
+        self.ptable_container = ipw.VBox(
+            children=(self.select_any_all, self.ptable),
+            layout={
+                "width": "auto",
+                "height": "auto" if extended else "0px",
+                "visibility": "visible" if extended else "hidden",
```
Comment (csadorf) [inline optimade_client/subwidgets/periodic_table.py:39]: I assume that you are passing those parameters on construction to avoid unwanted change of widget state on initialization? If yes, that's good.

**Review by csadorf — APPROVED** [review]

File: optimade_client/subwidgets/periodic_table.py (hunk 5 above)
Comment (CasperWA) [reply optimade_client/subwidgets/periodic_table.py:39]: Indeed. So if initialized with `extended=False` (i.e., one passes `embedded=True` to the wrapper widget) then it should make sure everything is initalized correctly as if the periodic table widget is hidden.

**Comment by CasperWA** [comment]
@csadorf I will rebase and merge some commits and merge. No need to re-approve.
